### PR TITLE
Removing trailing whitespace inside multiline strings is unsafe

### LINKED ISSUE #8037: W291: ruff --fix removes trailing whitespace in multiline string
When `W291` ([trailing-whitespace](https://docs.astral.sh/ruff/rules/trailing-whitespace/)) is enabled, ruff warns about trailing whitespace in multiline strings.
When `--fix` is passed, ruff modifies the multiline string.

## Repro:

```python
# repro.py
string = """
this line has trailing whitespace 
"""
```
```
$ ruff repro.py --isolated --select W291
repro.py:3:34: W291 [*] Trailing whitespace
Found 1 error.
[*] 1 potentially fixable with the --fix option.
$ ruff repro.py --isolated --select W291 --fix  # This modifies the multiline string!
Found 1 error (1 fixed, 0 remaining).
$ ruff --version
ruff 0.0.278
```

I noticed that the [implementation `trailing_whitespace`](https://github.com/astral-sh/ruff/blob/dda4ceda715c185aa1916cf361bbd31e1d28a9a0/crates/ruff_linter/src/rules/pycodestyle/rules/trailing_whitespace.rs#L75) categorizes the `--fix` action as [`Fix::safe_edit`](https://github.com/astral-sh/ruff/blob/dda4ceda715c185aa1916cf361bbd31e1d28a9a0/crates/ruff_linter/src/rules/pycodestyle/rules/trailing_whitespace.rs#L105C41-L105C41). I feel that modifying a multiline string is not safe; it can change the behavior of my python code. The doc for [`Applicability::Safe`](https://github.com/astral-sh/ruff/blob/dda4ceda715c185aa1916cf361bbd31e1d28a9a0/crates/ruff_diagnostics/src/fix.rs#L21-L23) say:
```
    /// The fix is safe and can always be applied.
    /// The fix is definitely what the user intended, or it maintains the exact meaning of the code.
```
I feel that this does not apply to `W291` as implemented.

### PR DESCRIPTION
**Pull request by sanxiyn**
Fix #8037.

### CODE REVIEW

File: crates/ruff_linter/resources/test/fixtures/pycodestyle/W291.py [hunk 1]
```diff
@@ -0,0 +1,2 @@
+'''trailing whitespace 
+inside a multiline string'''
```
Comment (charliermarsh) [inline crates/ruff_linter/resources/test/fixtures/pycodestyle/W291.py:2]: Do you mind adding a test for a multi-line f-string here?

**Review by charliermarsh — COMMENTED** [review]
Nice, thanks!

File: crates/ruff_linter/src/rules/pycodestyle/rules/trailing_whitespace.rs [hunk 2]
```diff
@@ -101,8 +101,16 @@ pub(crate) fn trailing_whitespace(
...
+            let safe = !indexer.multiline_ranges().intersects(range);
             let mut diagnostic = Diagnostic::new(TrailingWhitespace, range);
-            diagnostic.set_fix(Fix::safe_edit(Edit::range_deletion(range)));
+            let edit = Edit::range_deletion(range);
+            let fix = if safe {
+                Fix::safe_edit(edit)
+            } else {
+                Fix::unsafe_edit(edit)
+            };
+            diagnostic.set_fix(fix);
```
[the remark this answers is not in the document]
Comment (sanxiyn) [reply crates/ruff_linter/src/rules/pycodestyle/rules/trailing_whitespace.rs:113]: Done!

File: crates/ruff_linter/resources/test/fixtures/pycodestyle/W291.py (hunk 1 above)
Comment (sanxiyn) [reply crates/ruff_linter/resources/test/fixtures/pycodestyle/W291.py:2]: I added a test and it failed! From a brief look, it didn't seem trivial, and I would prefer to do this as a separate PR.

Comment (charliermarsh) [reply crates/ruff_linter/resources/test/fixtures/pycodestyle/W291.py:2]: Sounds good!

**Review by charliermarsh — APPROVED** [review]
Thanks!

**Pull request merged**